### PR TITLE
[issue 15229] Range-ified BigInt's Ctor to Accept Ranges of Characters

### PR DESCRIPTION
**Pull request by JackStouffer**
* ~~This contains the commit from https://github.com/D-Programming-Language/phobos/pull/3874 which this PR requires and was separated out for convenience~~
* ~~This is technically a breaking change because the ctor no longer errors on parsing problems, but throws~~
* The decimal path requires only a forward range, but the hex path requires a bidirectional range. I would like to make it possible for people to pass in forward ranges if they are just using decimals, as bidirectional ranges are less common than forward ranges. But, I couldn't figure out how to make the signiture accept any forward range and only accept bidriectional ranges for hexs at run time.
* For some reason I get a out of memory error on my machine for a specific unittest block. The error message is very unhelpful, so any guidance in this area would be appreciated.
* If said unittest block is commented out, all other tests pass.

### CODE REVIEW

**Comment by quickfur** [comment]
Autotester is failing with some pretty nasty errors. Implementation bug, perhaps?

**Comment by JackStouffer** [comment]
@quickfur See the last two bullet points

**Comment by quickfur** [comment]
Does it help if you split the unittest into smaller chunks? Or is the code interdependent?

**Comment by JackStouffer** [comment]
I will check that when I get the chance, but it makes me worried and I would like to know the root cause if possible.

**Comment by quickfur** [comment]
I wasn't suggesting that as a solution, but more a way to narrow down what might be going wrong.

**Comment by JackStouffer** [comment]
This assert is the culprit https://github.com/JackStouffer/phobos/blob/4f1d199538bbeaf47d14e700ae5915b3f9de8f69/std/bigint.d#L1151

commenting it out makes all tests pass. The only reason I can think of for it throwing an out of memory error is the `toString` function called via the `toHex` function in the test, as none of the rewritten functions allocate.

**Comment by quickfur** [comment]
Thanks for the tip, and sorry for the delay.... I managed to reproduce the problem locally; I should be able to narrow it down to learn why it's happening. Definitely something fishy is going on with the code here...

File: std/internal/math/biguintcore.d [hunk 1]
```diff
@@ -345,35 +346,47 @@ public:
...
+            s.popFront;
+        }
+
+        if (s.empty)
+        {
+            data = ZERO;
+            return true;
+        }
+
+        auto len = (s.save.walkLength - first_non_zero + 15) / 4;
```
Comment (quickfur) [inline std/internal/math/biguintcore.d]: This line is the cause of your bug.  In the `while` loop just before this, you already consumed all the leading zeroes, so `walkLength` no longer includes the zeroes in the count. However, you still subtract by `first_non_zero`. When the latter happens to be longer than the length of the remaining digits, a `size_t` underflow happens and wraps around to a very large value (to be precise, 4611686018427387903), which the following line then tries to allocate a buffer for.  So the out-of-memory error should not be surprising. :-)

The fix is trivial: there is no need to subtract `first_non_zero` at all, since you have already discarded those zeroes in the preceding `while` loop. Remove that, and everything should work as expected.

(Aren't you glad there's a unittest that just happens to have a test case that triggered this bug? It would have slipped in unnoticed otherwise...)

Comment (quickfur) [inline std/internal/math/biguintcore.d]: P.S. In fact, you probably don't even need to compute `first_non_zero`, since the zeroes are already consumed by the loop, and you can just work with the remaining digits directly.

File: std/internal/math/biguintcore.d [hunk 2]
```diff
@@ -395,26 +408,30 @@ public:
...
+            s.popFront;
         }
-        if (firstNonZero == s.length && s.length >= 1)
+
+        if (s.empty)
         {
             data = ZERO;
             return true;
         }
-        auto predictlength = (18*2 + 2*(s.length-firstNonZero)) / 19;
```
Comment (quickfur) [inline std/internal/math/biguintcore.d]: This line has the same bug. Unfortunately no unittest triggers the problem. (Add one?)

File: std/internal/math/biguintcore.d [hunk 3]
```diff
@@ -395,26 +408,30 @@ public:
...
-        auto tmp = new BigDigit[predictlength];
+        auto predict_length = (18 * 2 + 2 * (s.save.walkLength - first_non_zero)) / 19;
```
Comment (quickfur) [inline std/internal/math/biguintcore.d]: This line has the same bug. But no unittest caught it (add one?).

**Comment by JackStouffer** [comment]
Thank you for finding the root cause of the bug.

I have added a unit test for the decimal walk length bug that used to fail but with the latest changes passes. 

> Aren't you glad there's a unittest that just happens to have a test case that triggered this bug? It would have slipped in unnoticed otherwise

Thanks Don Clugston for your thorough unit tests!

**Comment by JackStouffer** [comment]
There is still the issue of the forward vs bidirectional range issue I made in the original post. Here is a good solution IMO:

* In existing code, all arguments passed to the Ctor are regular strings
* Because of that, I can add a new Ctor that only covers forward ranges (excludes bidirectional ranges) and throws if it detects a hex value, while leaving the bidirectional Ctor option available. 
* All existing code will still work because it will go to the bidirectional option, and only new code could target the forward range Ctor

Opinions?

File: std/internal/math/biguintcore.d [hunk 4]
```diff
@@ -395,26 +406,23 @@ public:
...
+        if (s.empty)
         {
             data = ZERO;
             return true;
         }
-        auto predictlength = (18*2 + 2*(s.length-firstNonZero)) / 19;
-        auto tmp = new BigDigit[predictlength];
 
-        uint hi = biguintFromDecimal(tmp, s[firstNonZero..$]);
+        auto predict_length = (18 * 2 + 2 * s.save.walkLength) / 19;
```
Comment (quickfur) [inline std/internal/math/biguintcore.d]: Why are leading zeroes no longer stripped?

**Comment by quickfur** [comment]
Hmm. In theory, there is no need to require bidirectional ranges for `fromHexString`. Sure, the current code needs it because of `foreach_reverse`, but since you're already using `walkLength`, it doesn't seem too much worse to also count the number of actual hex digits in the input range (i.e., ignore any `_`'s).

Once you know the actual number of hex digits, say let's call it `n`, then you can calculate the shift position required for the first digit (`n % ` *number of hex digits per word*), as well as which word in the `BigInt` it will end up in (`n / ` *number of hex digits per word*). So you can actually iterate from front to back (i.e., require only a forward range) by decrementing the shift position and moving to the previous word in the `BigInt`.

File: std/internal/math/biguintcore.d (hunk 4 above)
Comment (JackStouffer) [inline std/internal/math/biguintcore.d]: fixed

**Comment by quickfur** [comment]
P.S. in fact, counting actual hex digits (excluding `_`) will also give you a more accurate estimate of the number of words required to store the final value. So you could just replace `walkLength` with `count` using an appropriate predicate.

**Comment by quickfur** [comment]
Oh actually, just noticed that by the time you get to `biguintcore` you've already filtered out `_`. So you could just use `walkLength` to compute the shift and word number directly.

**Comment by JackStouffer** [comment]
Could you elaborate, or direct me to a resource that does already, on what you mean by "hex digits per word"?

**Comment by quickfur** [comment]
Basically, `BigInt` stores its value as an array of words (currently either `uint` or `ushort`). Each word stores either 4 or 8 hex digits. So, if you are given a hex literal containing, say, n=21 digits, and let's say we're using ushort (4 digits per word), that means the final result will occupy 6 words (24 digits, basically n/4 rounded up). The first digit in the literal will end up as the 3rd digit of the 6th word (i.e., the (n%4)'th digit of the (n/4)'th word). So once you know this, you can process the digits in order, without needing bidirectionality: start with the 3rd digit of the 6th word, convert the first digit, then the 2nd digit of the 6th word, convert the second digit, etc., then the 4th digit of the 5th word for the 4th digit, etc.. When you reach the last digit you should be on the 1st digit of the 1st word.

To store a converted digit into the i'th digit of the j'th word, you just do `tmp[j] |= (digit << (4*i))`.

So basically, the initial scan of the forward range will determine how many hex digits you have, then you compute which digit of which word corresponds with the first digit, and then scan the forward range the 2nd time, converting each digit and writing them into the result.

**Comment by JackStouffer** [comment]
@quickfur I don't have the time anymore to enhance this, but I think this is definitely an improvement over the current state of the code, so I think this should be pulled as is.

If this gets pulled, I will make an enhancement request on the issue tracker on the improvements that should be further made to this.

**Comment by quickfur** [comment]
ping random Phobos reviewers @andralex @yebblies @burner @DmitryOlshansky @schveiguy  
I don't have time to review this right now, but it's been sitting here for a long time. Can somebody else pick it up and see if it's merge-worthy?

File: std/bigint.d [hunk 5]
```diff
@@ -49,44 +50,113 @@ private:
...
+        auto filtered_range = s.filterBidirectional!(a => a != '_');
+        auto len = filtered_range.save.walkLength;
+        ubyte i;
+
+        data = 0UL;
+
+        // check for signs and if the string is a hex value
+        if (len > 1)
+        {
+            for (i = 0; i < 2; i++)
```
Comment (quickfur) [inline std/bigint.d]: I find this logic needlessly convoluted. Sign checking doesn't need to be interwoven with checking for `0x`. It should be a very simple, 2-step process:
````
// Check sign
if (range.front == '+')
    range.popFront(); // skip '+'
else if (range.front == '-')
{
    neg = true;
    range.popFront();
}

// Check for hex prefix
if (len > 2 && range.save.startsWith("0x"))
    // is hex
else
   // is decimal
````

Furthermore, filtering out `_` seems to be done prematurely. Do we really support literals of the form `_-_0_x_1234_567A`?? I would think that the negative sign and the `0x` prefix should not allow any intervening `_`'s. So really, sign checking and hex prefix checking should be done first, and *then* `_` gets filtered out.

File: std/bigint.d [hunk 6]
```diff
@@ -49,44 +50,113 @@ private:
...
-        } else {
-            ok = data.fromDecimalString(s);
+            ok = data.fromHexString(filtered_range);
         }
-        if (!ok)
-            throw new ConvException("Invalid digit string", file, line);
+        else
+        {
+            ok = data.fromDecimalString(buff[]
+                .filter!(a => a != ElementType!Range.init)
```
Comment (quickfur) [inline std/bigint.d]: What's this filter for?

File: std/internal/math/biguintcore.d [hunk 7]
```diff
@@ -345,35 +346,45 @@ public:
...
+                continue;
+
+            uint x = (character>='0' && character<='9') ? character - '0'
+                   : (character>='A' && character<='F') ? character - 'A' + 10
+                   : (character>='a' && character<='f') ? character - 'a' + 10
                    : 100;
-            if (x==100) return false;
+
+            if (x == 100)
+                return false;
```
Comment (quickfur) [inline std/internal/math/biguintcore.d]: This use of `100` as a nonsense value seems to be needless. Why not just break down the long-winded series of `?:`'s into a series of if-else statements ('cos that's what they get translated into anyway)? It will be easier to read and doesn't leave people wondering where on earth the `100` came from.

File: std/bigint.d (hunk 6 above)
Comment (JackStouffer) [inline std/bigint.d]: `buff[].chain(filtered_range)` produces garbage if one of the values wasn't set.

File: std/bigint.d (hunk 5 above)
Comment (JackStouffer) [inline std/bigint.d]: The docs do state "Underscores are permitted in any location", but that was a lie anyway, so I will update that.

**Comment by JackStouffer** [comment]
@quickfur Fixed

File: std/bigint.d [hunk 8]
```diff
@@ -49,44 +50,80 @@ private:
...
+        auto big1 = BigInt(r1);
+        assert(big1 == BigInt(101));
+
+        auto r2 = new ReferenceBidirectionalRange!dchar("1_000");
+        auto big2 = BigInt(r2);
+        assert(big2 == BigInt(1000));
+
+        auto r3 = new ReferenceBidirectionalRange!dchar("0x0");
+        auto big3 = BigInt(r3);
+        assert(big3 == BigInt(0));
```
Comment (quickfur) [inline std/bigint.d]: Did you test if "0x" is wrongly accepted to mean "0x0"?

**Comment by quickfur** [comment]
The rest looks good.  The `foreach_reverse` technically can be replaced so that we can relax the bidirectional requirement, but since that's a pure enhancement, I think we can merge this first and implement that in a follow-up PR.  LGTM once you add the check that "0x" is rejected.

**Comment by JackStouffer** [comment]
@quickfur Just to let you know because I am overwriting your changes, Walter is very against using __FILE__ and __LINE__ in exceptions: https://issues.dlang.org/show_bug.cgi?id=13543

**Comment by quickfur** [comment]
No problem, Walter is right. :-)

**Comment by JackStouffer** [comment]
@quickfur Fixed and added test case.

File: std/internal/math/biguintcore.d [hunk 9]
```diff
@@ -1598,14 +1621,19 @@ size_t biguintToDecimal(char [] buff, BigDigit [] data) pure nothrow
...
 in
 {
-    assert((data.length >= 2) || (data.length == 1 && s.length == 1));
+    static if (hasLength!Range)
+        assert((data.length >= 2) || (data.length == 1 && s.length == 1));
 }
 body
 {
     import std.conv : ConvException;
+    import std.stdio;
```
Comment (quickfur) [inline std/internal/math/biguintcore.d]: What's this import for?

Comment (JackStouffer) [inline std/internal/math/biguintcore.d]: Oops

**Comment by quickfur** [comment]
Looks ready to merge, other than the stray `import std.stdio` (that I assume is leftover from debugging writeln's?). Let's take that out and get this PR merged.

**Comment by JackStouffer** [comment]
@quickfur fixed.

Yeah, you have to use `writeln` on OS X, no debuggers work :/

**Comment by quickfur** [comment]
Strange, I still see the import in the diff?

I actually prefer using `writeln` for debugging, because at my day job I work with embedded software, and it's a pain to setup remote debugging and ensure gdb can find the libraries, source code, and symbols (any or all of which may be out of sync with what's actually running on the box), and it's a pain to hack the makefiles to (re!)compile with `-g` as well as suppress possible code-reordering caused by `-O`. All in all, a lot of pain just to be able to run a debugger, wastes more time than typing `printf` or `writeln`, recompile just that one module, and upload it to the box. Better yet, it helps with debugging code that runs when the box is booting and there's no way to actually start a debugger at that point in time -- just `fprintf` into a file (`tmpfile.writefln`) and examine the file after the box has started up.

**Comment by JackStouffer** [comment]
Sorry didn't force push.

File: std/internal/math/biguintcore.d [hunk 10]
```diff
@@ -1624,14 +1650,15 @@ body
     if (data.length > 1)
         data[1] = 0;
 
-    for (int i= (s[0]=='-' || s[0]=='+')? 1 : 0; i<s.length; ++i)
+    foreach (character; s)
     {
-        if (s[i] == '_')
+        //character.writeln(" char \n");
+        if (character == '_')
```
Comment (quickfur) [inline std/internal/math/biguintcore.d]: Hmm. Aren't all `_`'s already stripped by the time the code gets here?

File: std/internal/math/biguintcore.d [hunk 11]
```diff
@@ -1624,14 +1650,15 @@ body
     if (data.length > 1)
         data[1] = 0;
 
-    for (int i= (s[0]=='-' || s[0]=='+')? 1 : 0; i<s.length; ++i)
+    foreach (character; s)
     {
-        if (s[i] == '_')
+        //character.writeln(" char \n");
```
Comment (quickfur) [inline std/internal/math/biguintcore.d]: Also, this comment shouldn't be there. :-)

**Comment by JackStouffer** [comment]
@quickfur fixed

**Comment by quickfur** [comment]
Oh no, autotester is failling with "invalid digit" errors from biguintcore. :-(

**Comment by JackStouffer** [comment]
Works on my machine. Not quite sure what to do here.

**Comment by schveiguy** [comment]
1731 contains nothing that would cause an exception:
https://github.com/JackStouffer/phobos/blob/bigint2/std/internal/math/biguintcore.d#L1731

WTF?

**Comment by JackStouffer** [comment]
I kicked the auto-tester, so hopefully it will pass now.

**Comment by JackStouffer** [comment]
Nope, didn't work.

I'm out of ideas.

**Comment by JackStouffer** [comment]
@quickfur @schveiguy Added back in supposedly redundant check, it's green now.

**Comment by schveiguy** [comment]
64 bit tester now running out of memory. CC @braddr

**Comment by braddr** [comment]
Nothing wrong with the tester host itself, as far as I can tell.  Something's pushed the memory usage up over the threshold or close enough to it to fail periodically.

**Comment by JackStouffer** [comment]
Looks good now. Anything else need fixing?

**Comment by JackStouffer** [comment]
Ping @quickfur @schveiguy

**Comment by JackStouffer** [comment]
Ping. This has been sitting here ready to go for a while now.

File: std/bigint.d [hunk 12]
```diff
@@ -49,44 +50,88 @@ private:
...
     {
-        if (s.length == 0)
-            throw new ConvException("Can't initialize BigInt with "~
-                                    "empty string", file, line);
+        import std.algorithm.iteration : filterBidirectional;
+        import std.algorithm.searching : startsWith;
+        import std.exception : enforce;
+        import std.conv : ConvException;
+
+        enforce!ConvException(!s.empty, "Can't initialize BigInt with an empty range");
```
Comment (wilzbach) [inline std/bigint.d]: Can't we use assert for such simple checks?

File: std/bigint.d [hunk 13]
```diff
@@ -49,44 +50,88 @@ private:
...
+            s.popFront();
         }
-        data = 0UL;
-        bool ok;
-        assert(isZero());
-        if (s.length > 2 && (s[0..2] == "0x" || s[0..2] == "0X"))
+
+        if (s.save.startsWith("0x") || s.save.startsWith("0X"))
+        {
+            s.popFront; s.popFront;
```
Comment (wilzbach) [inline std/bigint.d]: Nitpick: shouldn't we have only one statement per line?

File: std/internal/math/biguintcore.d [hunk 14]
```diff
@@ -345,35 +346,55 @@ public:
...
-            assert(i>=0);
-            char c = s[i];
-            if (s[i]=='_') continue;
-            uint x = (c>='0' && c<='9') ? c - '0'
-                   : (c>='A' && c<='F') ? c - 'A' + 10
-                   : (c>='a' && c<='f') ? c - 'a' + 10
-                   : 100;
-            if (x==100) return false;
+        while (!s.empty && s.front == '0')
+        {
```
Comment (wilzbach) [inline std/internal/math/biguintcore.d]: Nitpick: No braces needed here

File: std/bigint.d (hunk 12 above)
Comment (JackStouffer) [inline std/bigint.d]: see https://issues.dlang.org/show_bug.cgi?id=15678

**Comment by JackStouffer** [comment]
@wilzbach Fixed

File: std/internal/math/biguintcore.d [hunk 15]
```diff
@@ -395,26 +414,27 @@ public:
         return true;
     }
 
-    // return true if OK; false if erroneous characters found
-    // FIXME: actually throws `ConvException` on error.
-    bool fromDecimalString(const(char)[] s) pure @trusted
+    /// return true if OK; false if erroneous characters found
```
Comment (DmitryOlshansky) [inline std/internal/math/biguintcore.d]: Should not be documented?

File: std/internal/math/biguintcore.d [hunk 16]
```diff
@@ -1624,14 +1648,15 @@ body
     if (data.length > 1)
         data[1] = 0;
 
-    for (int i= (s[0]=='-' || s[0]=='+')? 1 : 0; i<s.length; ++i)
```
Comment (DmitryOlshansky) [inline std/internal/math/biguintcore.d]: Where did that check go? (i.e. skipping of '-' and '+')

Comment (JackStouffer) [inline std/internal/math/biguintcore.d]: By the time the string gets here, the sign is already popped from the range, see https://github.com/JackStouffer/phobos/blob/bigint2/std/bigint.d#L82

**Comment by JackStouffer** [comment]
@DmitryOlshansky Fixed

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled on

**Comment by wilzbach** [comment]
Can we change the title so that the dlang bot finds this? :)

File: std/internal/math/biguintcore.d [hunk 17]
```diff
@@ -345,35 +346,53 @@ public:
...
+        while (!s.empty && s.front == '0')
+            s.popFront;
+
+        if (s.empty)
+        {
+            data = ZERO;
+            return true;
+        }
+
+        auto len = (s.save.walkLength + 15) / 4;
```
Comment (wilzbach) [inline std/internal/math/biguintcore.d]: May I ask why we have the magic `+15`? If I read the below correctly `(s.length/ 8) + 1` should work as we use at most 1/8 of tmp per char.

Comment (DmitryOlshansky) [inline std/internal/math/biguintcore.d]: The usual trick of 16-1, so it rounds up correctly when divided by 16. However it is then multiplied by 4 so net total is /4

Comment (wilzbach) [inline std/internal/math/biguintcore.d]: > multiplied by 4 so net total is /4

Imho it's `8` - see:

```
++partcount;
if (partcount == 8)
```

File: std/internal/math/biguintcore.d [hunk 18]
```diff
@@ -345,35 +346,53 @@ public:
...
+                return false;
+            }
+
             part >>= 4;
-            part |= (x<<(32-4));
+            part |= (x << (32 - 4));
             ++partcount;
-            if (partcount==8)
+
+            if (partcount == 8)
```
Comment (wilzbach) [inline std/internal/math/biguintcore.d]: shouldn't we do a module or (better) reset?

**Comment by JackStouffer** [comment]
Thanks!

**Comment by aG0aep6G** [comment]
This has caused a regression. `immutable(BigInt)("123")` doesn't work anymore. In Bugzilla: [issue 17330](https://issues.dlang.org/show_bug.cgi?id=17330).
